reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/FlatMapTest.kt

```diff
@@ -0,0 +1,193 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.testutils.TestObservable
+import com.badoo.reaktive.testutils.TestObservableObserver
+import com.badoo.reaktive.testutils.dispose
+import com.badoo.reaktive.testutils.hasOnNext
+import com.badoo.reaktive.testutils.hasSubscribers
+import com.badoo.reaktive.testutils.isCompleted
+import com.badoo.reaktive.testutils.isError
+import com.badoo.reaktive.testutils.test
+import com.badoo.reaktive.testutils.values
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class FlatMapTest : UpstreamDownstreamGenericTests by UpstreamDownstreamGenericTests<Unit>({ flatMap { TestObservable<Int>() } }) {
+
+    private val source = TestObservable<Int>()
+
+    @Test
+    fun subscribes_to_upstream() {
+        testFlatMap { TestObservable() }
+
+        assertTrue(source.hasSubscribers)
+    }
+
+    @Test
+    fun subscribes_to_inner_sources() {
+        val inners = listOf<TestObservable<String>>(TestObservable(), TestObservable())
+        testFlatMap(inners::get)
+
+        source.onNext(0)
+        source.onNext(1)
+
+        assertTrue(inners[0].hasSubscribers)
+        assertTrue(inners[1].hasSubscribers)
+    }
+
+    @Test
+    fun produces_values_in_correct_order() {
+        val inners = listOf<TestObservable<String>>(TestObservable(), TestObservable(), TestObservable())
+        val observer = testFlatMap(inners::get)
+
+        source.onNext(0)
+        inners[0].onNext("0a")
+        source.onNext(1)
+        inners[0].onNext("0b")
+        inners[1].onNext("1a")
+        source.onNext(2)
+        inners[1].onNext("1b")
+        inners[2].onNext("2a")
+        inners[2].onNext("2b")
+
+        assertEquals(listOf("0a", "0b", "1a", "1b", "2a", "2b"), observer.values)
+    }
+
+    @Test
+    fun completes_WHEN_upstream_completed_without_values() {
+        val observer = testFlatMap { TestObservable() }
+
+        source.onComplete()
+
+        assertTrue(observer.isCompleted)
+    }
+
+    @Test
+    fun does_not_complete_WHEN_upstream_produced_values_and_completed() {
```

Comment:
It does not complete, because stream to which it was flatMapped did not complete? Otherwise it should complete: `Observable.just(0).flatMap { Observable.just("$it") }` completes with value.

Reply:
Upstream is completed, but inner (flatmapped) stream is not